### DISCUSSION TITLE
C# keyword 'empty' compiles to the static Empty property of List, Dictionary, etc.

### DISCUSSION THREAD

**Discussion by JustinMinnaar** [dotnet/roslyn#66964] [Ideas] [answered]
Typically, when a function generates a list, instead of returning a NULL, it has been proposed as a best-practice to return a static empty list, preventing memory use for the creation of a new empty list, and allowing the caller to interate the list without concern about NULL. Perhaps a new keyword 'empty' could return the static property of the class with the name 'Empty'. As a convention, this may make this process easier to read.
```
public List<Person> GetPeople(PeopleFilter filter)
{
    if (someCondition)
        return new List<Person>();

    ...
}
```

This could be changed to 
```
public List<Person> GetPeople(PeopleFilter filter)
{
    if (someCondition)
        return empty;

    ...
}
```

Of course, this might only work with types that specifically implement a property for Empty, such as dictionary, list, etc.

**Comment by CyrusNajmabadi** [comment, marked as answer]
The collection-literals proposal (such were working on for the next version of c#) allows you to just use `[]` to represent 'empty' in all these cases.

**Comment by JustinMinnaar** [comment]
Does the collection-literals proposal use a single static instance of the empty collection, to prevent allocation memory on the heap for a new collection each time an empty is returned?

**Reply by CyrusNajmabadi** [reply]
No.  As that would be broken.  If you had a `List<Person>`  and it was a single static instance, then anyone could mutate it and would affect all other empty lists of that type :)

**Reply by JustinMinnaar** [reply]
That is an interesting thought. Can one mutate the existing `Array.Empty<string>()` result? Would we not just add a similar method, a new `List.Empty<string>()` method?

**Reply by CyrusNajmabadi** [reply]
No. You cannot add to arrays.  But you can certainly add to lists. That's the defining difference between the two.

**Reply by saint4eva** [reply]
https://www.youtube.com/watch?v=xIA3t2Je3uQ

**Reply by KennethHoff** [reply]
Does this mean that the collection literals will lower to something akin to this?

```c#
List<int> intList = []; // new List<int>(0);
int[] intArray = []; // Array.Empty<int>();
Span<int> intSpan = []; // stackalloc Array.Empty<int>(); // This might be entirely wrong. Never used a Span 🤷 
```

**Reply by CyrusNajmabadi** [reply]
Yes
